x/sync/syncmanager.go

```diff
@@ -361,93 +361,135 @@ func (m *StateSyncManager) getAndApplyRangeProof(ctx context.Context, workItem *
+
+		// select the deepest proof node from the two proofs
+		switch {
+		case receivedProofNode.KeyPath.NibbleLength > localProofNode.KeyPath.NibbleLength:
+			// there was a branch node in the received proof that isn't in the local proof
+			// see if the received proof node has children not present in the local proof
+			deepestNode = &receivedProofNode
+
+			// we have dealt with this received node, so move on to the next received node
+			receivedProofNodeIndex--
+
+		case localProofNode.KeyPath.NibbleLength > receivedProofNode.KeyPath.NibbleLength:
+			// there was a branch node in the local proof that isn't in the received proof
+			// see if the local proof node has children not present in the received proof
+			deepestNode = &localProofNode
+
+			// we have dealt with this local node, so move on to the next local node
+			localProofNodeIndex--
+
+		default:
+			// the two nodes are at the same depth
+			// see if any of the children present in the local proof node are different
+			// from the children in the received proof node
+			deepestNode = &localProofNode
+			deepestNodeFromOtherProof = &receivedProofNode
+
+			// we have dealt with this local node and received node, so move on to the next nodes
+			localProofNodeIndex--
+			receivedProofNodeIndex--
 		}
 
-		if startKeyPath.NibbleLength <= branchNode.KeyPath.NibbleLength {
-			return append(start, 0), nil
+		// We only want to look at the children with keys greater than the proofKey.
+		// The proof key has the deepest node's key as a prefix,
+		// so only the next nibble of the proof key needs to be considered.
+		// If the deepest node has the same key as the key being proven,
+		// then all of its children have keys greater than the proof key, so we can start at byte 0
+		startingChildNibble := byte(0)
+
+		// If the deepest node has a key shorter than the key being proven,
+		// so look at the next nibble of the proof key to determine which children have larger keys
+		if deepestNode.KeyPath.NibbleLength < proofKeyPath.NibbleLength {
+			startingChildNibble = proofKeyPath.NibbleVal(deepestNode.KeyPath.NibbleLength) + 1
 		}
 
-		// the two nodes have different paths, so find where they branched
-		for nextKeyNibble := startKeyPath.NibbleVal(branchNode.KeyPath.NibbleLength) + 1; nextKeyNibble < 16; nextKeyNibble++ {
-			if _, ok := branchNode.Children[nextKeyNibble]; ok {
-				result = branchNode.KeyPath.AppendNibble(nextKeyNibble).Value
-				break
-			}
+		if childIndex, hasDifference := findChildDifference(deepestNode, deepestNodeFromOtherProof, startingChildNibble); hasDifference {
+			nextKey = deepestNode.KeyPath.AppendNibble(childIndex).Value
+			break
 		}
 	}
 
-	if result == nil || (len(end) > 0 && bytes.Compare(result, end) >= 0) {
+	// if the result is before or equal to the lastReceivedKey
+	// then use the lastReceivedKey + 0, which is the next largest key after lastReceivedKey
+	if nextKey != nil && bytes.Compare(nextKey, lastReceivedKey) <= 0 {
```

Comment:
Not sure if this is required, but it isn't an expensive safe guard, so I think it is worth keeping.